### PR TITLE
Add methods and type for `MatrixSets`

### PR DESCRIPTION
**Pull request by alecarraro**
This is the first of a sequence of 3/4 pull requests to implement the overapproximation methods for maps with matrix zonotopes. In this PR i implement 

- An `AbstractMatrixZonotope` supertype
- A `MatrixZonotopeProduct` type to store lazy products of matrix zonotopes
- A MatrixZonotopeExp` to represent lazily the expoenntial $exp(\mathcal{A})$
- norm computation and overapproximations for `MatrixZonotopeProduct`
- linear map for matrix zonotope

### CODE REVIEW

**Comment by schillic** [comment]
Could you add a separate PR for fixing `scale`, including a test?

**Comment by alecarraro** [comment]
> Could you add a separate PR for fixing `scale`, including a test?

Done. Should I undo the commit in this one then?

**Comment by schillic** [comment]
Yes - there will be a merge conflict sooner or later anyway.

**Comment by schillic** [comment]
Still a conflict here...

File: src/MatrixSets/scale.jl [hunk 1]
```diff
@@ -9,3 +9,10 @@ function scale!(α::Real, MZ::MatrixZonotope)
     end
     return MZ
 end
+
+"""
+    *(a::Real, B::MatrixZonotope)
+
+Alias to scale a matrix zonotope.
+"""
+@commutative Base.:*(α::Real, MZ::MatrixZonotope) = scale(α, MZ)
```
Comment (schillic) [inline src/MatrixSets/scale.jl]: Since we said that `*` should not be used to multiply a matrix with a MZ, we should also not make it an alias for `scale` for consistency.

File: test/MatrixSets/MatrixZonotope.jl [hunk 2]
```diff
@@ -1,4 +1,5 @@
 using LazySets, Test
+import LazySets.MatrixZonotopeModule: factors
```
Comment (schillic) [inline test/MatrixSets/MatrixZonotope.jl]:
```suggestion
```
This should not be needed because `factors` is exported.

Also, it should be `using`: `import` of functions is used to extend them. I know it is confusing because `import` of packages has a different meaning.

File: src/MatrixSets/scale.jl (hunk 1 above)
Comment (alecarraro) [reply src/MatrixSets/scale.jl]: I agree that it is not consistent. In the `overapproximate` functions this is used quite a lot tho, and i think having to write `scale` every time is overly verbose.  

So I kinda changed my mind and now I believe that we could also have `*` as an alias for the `linear_map` since, as you said, in this context the meaning can be well understood. 

If you still think it is a good idea, I will add it back and add a note in the `linear_map` docstring, otherwise I'll remove this alias from `scale` as well

Comment (schillic) [reply src/MatrixSets/scale.jl]: I prefer consistency, and hence `*` should not be an alias for a concrete operation. `*` is also defined between matrix sets to be a lazy operation. So I suggest to stick to `scale`.

File: test/MatrixSets/MatrixZonotope.jl (hunk 2 above)
Comment (schillic) [reply test/MatrixSets/MatrixZonotope.jl]: It seems this was needed for disambiguation with `Symbolics`. So you have to re-add it as:
```julia
using LazySets.MatrixZonotopeModule: factors
```

Comment (alecarraro) [reply test/MatrixSets/MatrixZonotope.jl]: ah yes, now I rememember why I added it

File: docs/src/lib/matrixsets/MatrixZonotope.md [hunk 3]
```diff
@@ -29,3 +41,39 @@ Undocumented implementations:
 * [`eltype`](@ref eltype(::Type{<:LazySet}))
 * [`scale`](@ref scale(::Real, ::LazySet))
 * [`scale!`](@ref scale!(::Real, ::LazySet))
+* [`eltype`](@ref eltype(::Type{<:LazySet}))
```
Comment (schillic) [inline docs/src/lib/matrixsets/MatrixZonotope.md]:
```suggestion
```
This was already included three lines above.

File: docs/src/lib/matrixsets/MatrixZonotope.md [hunk 4]
```diff
@@ -29,3 +41,39 @@ Undocumented implementations:
...
+
+```@meta
+CurrentModule = LazySets.MatrixZonotopeModule
+```
+
+Inherited from [`AbstractMatrixZonotope`](@ref):
```
Comment (schillic) [inline docs/src/lib/matrixsets/MatrixZonotope.md]:
```suggestion
Undocumented implementations:
```
This is not inherited.

File: docs/src/lib/matrixsets/MatrixZonotope.md [hunk 5]
```diff
@@ -29,3 +41,39 @@ Undocumented implementations:
...
+* [`size`](@ref size(::AbstractMatrixZonotope))
+
+```@meta
+CurrentModule = LazySets.MatrixZonotopeModule
+```
+
```
Comment (schillic) [inline docs/src/lib/matrixsets/MatrixZonotope.md]:
```suggestion
```
Not needed.

File: docs/src/lib/matrixsets/MatrixZonotope.md [hunk 6]
```diff
@@ -29,3 +41,39 @@ Undocumented implementations:
...
+MatrixZonotopeProduct
+factors
+nfactors
+remove_redundant_factors
+```
+
+```@meta
+CurrentModule = LazySets.MatrixZonotopeModule
+```
+
```
Comment (schillic) [inline docs/src/lib/matrixsets/MatrixZonotope.md]:
```suggestion
```

File: docs/src/lib/matrixsets/MatrixZonotope.md [hunk 7]
```diff
@@ -29,3 +41,39 @@ Undocumented implementations:
...
+## Operations 
+Inherited from [`AbstractMatrixZonotope`](@ref):
```
Comment (schillic) [inline docs/src/lib/matrixsets/MatrixZonotope.md]:
```suggestion
Undocumented implementations:
```

File: docs/src/lib/matrixsets/MatrixZonotope.md [hunk 8]
```diff
@@ -29,3 +41,39 @@ Undocumented implementations:
...
+* [`size`](@ref size(::AbstractMatrixZonotope))
+
+# [MatrixZonotopeExp](@id def_MatrixZonotopeExp)
+```@docs
+MatrixZonotopeExp
+```
+## Operations 
+Inherited from [`AbstractMatrixZonotope`](@ref):
```
Comment (schillic) [inline docs/src/lib/matrixsets/MatrixZonotope.md]:
```suggestion
Undocumented implementations:
```

File: src/Approximations/overapproximate_norm.jl [hunk 9]
```diff
@@ -103,3 +103,21 @@ function overapproximate_norm(MZ::MatrixZonotope, p::Real=Inf)
         throw(ArgumentError("the norm for p=$p has not been implemented"))
     end
 end
+
+"""
+    overapproximate_norm(MZ::MatrixZonotopeProduct, [p]::Real=Inf)
```
Comment (schillic) [inline src/Approximations/overapproximate_norm.jl]:
```suggestion
    overapproximate_norm(MZP::MatrixZonotopeProduct, [p]::Real=Inf)
```

File: src/MatrixSets/AbstractMatrixZonotope.jl [hunk 10]
```diff
@@ -0,0 +1,17 @@
+"""
+    AbstractMatrixZonotope{N}
+
+Abstract supertype for all matrix zonotope representations.
+
+Every concrete `AbstractSparsePolynomialZonotope` must define the following functions:
+- `size(::AbstractSparsePolynomialZonotope)` -- return the size of the matrix zonotope
```
Comment (schillic) [inline src/MatrixSets/AbstractMatrixZonotope.jl]:
```suggestion
- `size(::AbstractSparsePolynomialZonotope, [dim])` -- return the size of the matrix zonotope
```

File: src/MatrixSets/MatrixZonotopeModule.jl [hunk 11]
```diff
@@ -4,24 +4,27 @@ using Reexport
 
 using Random: AbstractRNG, GLOBAL_RNG
 using ReachabilityBase.Distribution: reseed!
-using ..LazySets.SparsePolynomialZonotopeModule
+using ReachabilityBase.Commutative
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeModule.jl]:
```suggestion
```
Not needed anymore

File: src/MatrixSets/MatrixZonotopeProduct.jl [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    return length(factors(MZP))
+end
+
+Base.size(P::MatrixZonotopeProduct) = (size(P.factors[1].A0, 1), size(P.factors[end].A0, 2))
+Base.size(P::MatrixZonotopeProduct, d::Int) = size(P)[d]
+
+"""
+    remove_redundant_factors(MZP::MatrixZonotopeProduct)
+
+Returns a simplified `MatrixZonotopeProduct` where generator-free factors are absorbed
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeProduct.jl]:
```suggestion
Return a simplified `MatrixZonotopeProduct` where generator-free factors are absorbed
```

File: src/MatrixSets/MatrixZonotopeProduct.jl [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+into adjacent factors via linear maps.
+
+### Input
+
+- `MZP` -- matrix zonotope product.
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeProduct.jl]:
```suggestion
- `MZP` -- matrix zonotope product
```

File: src/Sets/SparsePolynomialZonotope/SparsePolynomialZonotopeModule.jl [hunk 14]
```diff
@@ -6,20 +6,20 @@ using ..LazySets: AbstractSparsePolynomialZonotope, AbstractReductionMethod,
                   genmat, GIR05, order, _remove_redundant_generators_polyzono,
                   @validate
 import IntervalArithmetic as IA
+using LinearAlgebra: I
```
Comment (schillic) [inline src/Sets/SparsePolynomialZonotope/SparsePolynomialZonotopeModule.jl]:
```suggestion
```
I think this is not used anywhere here...

File: src/MatrixSets/linear_map.jl [hunk 15]
```diff
@@ -0,0 +1,37 @@
+"""
+    linear_map(M::AbstractMatrix, MZ::MatrixZonotope)
+
+Applies a linear transformation to a matrix zonotope from the left.
```
Comment (schillic) [inline src/MatrixSets/linear_map.jl]:
```suggestion
Apply a linear transformation to a matrix zonotope from the left.
```

File: src/MatrixSets/linear_map.jl [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+function linear_map(M::AbstractMatrix, MZ::MatrixZonotope)
+    @assert size(M, 2) == size(MZ, 1) "incompatible dimensions"
+    gens = [M * Ai for Ai in generators(MZ)]
+    return MatrixZonotope(M * center(MZ), gens)
+end
+
+"""
+    linear_map(M::AbstractMatrix, MZ::MatrixZonotope)
+
+Applies a linear transformation to a matrix zonotope from the right.
```
Comment (schillic) [inline src/MatrixSets/linear_map.jl]:
```suggestion
Apply a linear transformation to a matrix zonotope from the right.
```

File: test/MatrixSets/MatrixZonotope.jl [hunk 17]
```diff
@@ -60,6 +61,61 @@ for N in @tN([Float64, Float32, Rational{Int}])
...
+
+    # size checks
+    @test size(MZP) == (2, 3)
+    @test size(MZP, 1) == 2
+    @test size(MZP, 2) == 3
+
+    # chainable multiplication
+    MZP_chain = A_mz * B_mz * MatrixZonotope(N[1 0; 0 1; 1 1], [N[1 1; 0 0; 0 -2], N[0 1; 1 0; 1 -1]])
+    @test nfactors(MZP_chain) == 3
+    @test MZP_chain isa MatrixZonotopeProduct{N}
```
Comment (schillic) [inline test/MatrixSets/MatrixZonotope.jl:108]:
```suggestion
    @test MZP_chain isa MatrixZonotopeProduct{N}
    @test A_mz * B_mz == MZP
    @test A_mz * MZP == MatrixZonotopeProduct(A_mz, A_mz, B_mz)
    @test MZP * B_mz == MatrixZonotopeProduct(A_mz, B_mz, B_mz)
    @test MZP * MZP == MatrixZonotopeProduct(A_mz, B_mz, A_mz, B_mz)
```

File: test/MatrixSets/MatrixZonotope.jl [hunk 18]
```diff
@@ -60,6 +61,61 @@ for N in @tN([Float64, Float32, Rational{Int}])
...
+
+    # remove_redundant_factors
+    const_A = MatrixZonotope(N[1 0; 0 1], Vector{Matrix{N}}())
+    mixed = MatrixZonotopeProduct(const_A, A_mz, B_mz)
```
Comment (schillic) [inline test/MatrixSets/MatrixZonotope.jl]:
```suggestion
    const_B = MatrixZonotope(N[1 0 0; 0 1 0; 0 0 1], Vector{Matrix{N}}())
    mixed = MatrixZonotopeProduct(const_A, const_A, A_mz, const_A, const_A, B_mz, const_B, const_B)
```
To also test other corner cases.

File: test/MatrixSets/MatrixZonotope.jl [hunk 19]
```diff
@@ -60,6 +61,61 @@ for N in @tN([Float64, Float32, Rational{Int}])
...
+    simplified = remove_redundant_factors(mixed)
+    @test nfactors(simplified) == 2
+    @test size(simplified) == (2, 3)
```
Comment (schillic) [inline test/MatrixSets/MatrixZonotope.jl]:
```suggestion
    @test simplified == MZP
```

File: src/MatrixSets/MatrixZonotopeProduct.jl [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+### Output
+
+A matrix zonotope product with redundant constant factors removed.
+"""
+function remove_redundant_factors(MZP::MatrixZonotopeProduct)
+    factors_ = factors(MZP)
+    reduced = MatrixZonotope{eltype(factors_[1]), typeof(center(factors_[1]))}[]
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeProduct.jl]:
```suggestion
function remove_redundant_factors(MZP::MatrixZonotopeProduct{N,S}) where {N,S}
    factors_ = factors(MZP)
    reduced = MatrixZonotope{N,S}[]
```

File: src/MatrixSets/MatrixZonotopeProduct.jl [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+    i = 1
+    while i < length(factors_)
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeProduct.jl]: This can be turned into a normal `for` loop. Is that intended? See also the next comment.

File: src/MatrixSets/MatrixZonotopeProduct.jl [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+        if isempty(generators(MZ))
+            factors_[i+1] =  linear_map(center(MZ), factors_[i + 1])
+        else
+            push!(reduced, MZ)
+        end
+        i += 1
+    end
+
+    # in the last element apply linear map to the left
+    if i == length(factors_)
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeProduct.jl]: This condition always holds.

File: src/MatrixSets/MatrixZonotopeProduct.jl [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+```math
+\\mathcal{C} = \\{ A_1 A_2 \\dots A_n ~|~ A_i \\in \\mathcal{A}_i \\}
+```
+"""
+struct MatrixZonotopeProduct{N,S<:AbstractMatrix{N}} <: AbstractMatrixZonotope{N}
+    factors::Vector{MatrixZonotope{N,S}}
+
+    function MatrixZonotopeProduct(factors::Vector{<:MatrixZonotope{N,T}}) where {N,
+                                                                                  T<:AbstractMatrix{N}}
+        @assert length(factors) >= 2 "need at least two matrix zonotope factors"
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeProduct.jl]: Is there a good reason why the length cannot be 1? That would allow to get type-stable code.

File: src/MatrixSets/MatrixZonotopeProduct.jl [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+    if i == length(factors_)
+        last = factors_[end]
+        if isempty(generators(last)) && !isempty(reduced)
+            reduced[end] = linear_map(reduced[end], center(last))
+        else
+            push!(reduced, last)
+        end
+    end
+    if isempty(reduced)
+        return center(factors_[end-1]) # return concrete matrix
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeProduct.jl]: Why `end-1`?

File: src/MatrixSets/MatrixZonotopeProduct.jl [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+"""
+function remove_redundant_factors(MZP::MatrixZonotopeProduct)
+    factors_ = factors(MZP)
+    reduced = MatrixZonotope{eltype(factors_[1]), typeof(center(factors_[1]))}[]
+
+    i = 1
+    while i < length(factors_)
+        MZ = factors_[i]
+        if isempty(generators(MZ))
+            factors_[i+1] =  linear_map(center(MZ), factors_[i + 1])
```
Comment (schillic) [inline src/MatrixSets/MatrixZonotopeProduct.jl]: I do not understand this code. This function should not modify `factors_`.

File: src/Sets/SparsePolynomialZonotope/SparsePolynomialZonotopeModule.jl (hunk 14 above)
Comment (alecarraro) [reply src/Sets/SparsePolynomialZonotope/SparsePolynomialZonotopeModule.jl]: It will be used later for the linear map of matrix zonotope with spz. 
When i cherrypicked the commit to split the PR, I didnt notice that I left it here. I can remove it now , I will add it in the next PR

Comment (schillic) [reply src/Sets/SparsePolynomialZonotope/SparsePolynomialZonotopeModule.jl]: Okay, better remove it for now. Maybe we find out that it is not needed later. But not super important.

File: src/MatrixSets/MatrixZonotopeProduct.jl (hunk 24 above)
Comment (alecarraro) [reply src/MatrixSets/MatrixZonotopeProduct.jl]: ok, this was added to fix an edge case, but it actually breaks others. I will edit it and add more corner tests

**Comment by schillic** [comment]
Apart from `remove_redundant_factors` I can approve this PR. I have the feeling that `remove_redundant_factors` will take a bit longer to discuss. Do you want to move it to a separate PR so we can merge this? (Up to you.)

File: src/MatrixSets/MatrixZonotopeProduct.jl (hunk 23 above)
Comment (alecarraro) [reply src/MatrixSets/MatrixZonotopeProduct.jl]: Then I will do that. 
To make it type stable I could extend the constructor to support a single factor printing a warning.
And then iin case  all factors in a product have empty generators I return a `MatrixZonotopeProduct` with one factor with no generator and 2 warnings.
What do you think?

**Comment by alecarraro** [comment]
yes lets move it to a different PR,

File: src/MatrixSets/MatrixZonotopeProduct.jl (hunk 23 above)
Comment (schillic) [reply src/MatrixSets/MatrixZonotopeProduct.jl]: Why the warnings? `LazySets` lets you work with array sets of length 1 without warnings. Often the code automatically works and this special case does not require treatment.

Comment (schillic) [reply src/MatrixSets/MatrixZonotopeProduct.jl]: Maybe for this PR we remove the assertion and add it back later if we see that there are good reasons.
